lib/Frontend/ParseableInterfaceModuleLoader.cpp

```diff
@@ -0,0 +1,1022 @@
+                            DependencyTracker *tracker = nullptr)
+  : ctx(ctx), fs(*ctx.SourceMgr.getFileSystem()), diags(ctx.Diags),
+  interfacePath(interfacePath), moduleName(moduleName),
+  moduleCachePath(moduleCachePath), prebuiltCachePath(prebuiltCachePath),
+  serializeDependencyHashes(serializeDependencyHashes),
+  diagnosticLoc(diagnosticLoc), dependencyTracker(tracker) {
+    configureSubInvocation();
+  }
+
+  const CompilerInvocation &getSubInvocation() const {
+    return subInvocation;
+  }
+
+  bool buildSwiftModule(StringRef OutPath, bool ShouldSerializeDeps) {
+    bool SubError = false;
+    bool RunSuccess = llvm::CrashRecoveryContext().RunSafelyOnThread([&] {
+      // Note that we don't assume cachePath is the same as the Clang
+      // module cache path at this point.
+      if (!moduleCachePath.empty())
+        (void)llvm::sys::fs::create_directory(moduleCachePath);
+
+      configureSubInvocationInputsAndOutputs(OutPath);
+
+      FrontendOptions &FEOpts = subInvocation.getFrontendOptions();
+      const auto &InputInfo = FEOpts.InputsAndOutputs.firstInput();
+      StringRef InPath = InputInfo.file();
+      const auto &OutputInfo =
+        InputInfo.getPrimarySpecificPaths().SupplementaryOutputs;
+      StringRef OutPath = OutputInfo.ModuleOutputPath;
+
+      llvm::BumpPtrAllocator SubArgsAlloc;
+      llvm::StringSaver SubArgSaver(SubArgsAlloc);
+      SmallVector<const char *, 16> SubArgs;
+      swift::version::Version Vers;
+      if (extractSwiftInterfaceVersionAndArgs(Vers, SubArgSaver, SubArgs)) {
+        SubError = true;
+        return;
+      }
+
+      // For now: we support anything with the same "major version" and assume
+      // minor versions might be interesting for debugging, or special-casing a
+      // compatible field variant.
+      if (Vers.asMajorVersion() != InterfaceFormatVersion.asMajorVersion()) {
+        diags.diagnose(diagnosticLoc,
+                       diag::unsupported_version_of_parseable_interface,
+                       interfacePath, Vers);
+        SubError = true;
+        return;
+      }
+
+      SmallString<32> ExpectedModuleName = subInvocation.getModuleName();
+      if (subInvocation.parseArgs(SubArgs, diags)) {
+        SubError = true;
+        return;
+      }
+
+      if (subInvocation.getModuleName() != ExpectedModuleName) {
+        auto DiagKind = diag::serialization_name_mismatch;
+        if (subInvocation.getLangOptions().DebuggerSupport)
+          DiagKind = diag::serialization_name_mismatch_repl;
+        diags.diagnose(diagnosticLoc, DiagKind, subInvocation.getModuleName(),
+                       ExpectedModuleName);
+        SubError = true;
+        return;
+      }
+
+      // Optimize emitted modules. This has to happen after we parse arguments,
+      // because parseSILOpts would override the current optimization mode.
+      subInvocation.getSILOptions().OptMode = OptimizationMode::ForSpeed;
```

Comment:
I've been meaning to change this, since we now preserve the optimization level used to build clients, but I want to discuss that with Jordan when he gets back. rdar://46358840